dvc/api.py

```diff
@@ -1,15 +1,28 @@
+from contextlib import contextmanager
+
+from dvc.utils.compat import urlparse
 from dvc.repo import Repo
+from dvc.pkg.pkg import temp_repo
 
 
 def get_url(path, repo_dir=None, remote=None):
     """Returns an url of `path` in default or specified remote"""
-    repo = Repo(repo_dir)
-    out, = repo.find_outs_by_path(path)
-    remote_obj = repo.cloud.get_remote(remote)
-    return str(remote_obj.checksum_to_path_info(out.checksum))
+    with _make_repo(repo_dir) as repo:
+        out, = repo.find_outs_by_path(path)
+        remote_obj = repo.cloud.get_remote(remote)
+        return str(remote_obj.checksum_to_path_info(out.checksum))
 
 
 def open(path, repo_dir=None, remote=None, mode="r", encoding=None):
     """Opens a specified resource as a file descriptor"""
-    repo = Repo(repo_dir)
-    return repo.open(path, remote=remote, mode=mode, encoding=encoding)
+    with _make_repo(repo_dir) as repo:
+        return repo.open(path, remote=remote, mode=mode, encoding=encoding)
```

Comment:
1. repo.open implementation - won't work on a lot of NFS systems - we should find a way in these simple cases (pulling a single file) to avoid using DVC lock file and SQLlite.
2. repo.open implementation - it pulls the whole DVC file as far as I understand. All outputs. Sometimes it might mean a lot of unnecessary data.
3. It calls `clould.pull` - does it mean that it instantiates the data file in the temp workspace? By default we use copy semantics for our cache and we should be careful to avoid duplicating data.
3. I would double check if it will be possible to support file interface when we decide to do "so called" streaming. For example, for S3 we would need to use S3 fs or something? Or if we are not writing anything back it's not a big deal to wrap the byte stream.

Reply:
1  - there is a ticket for this already https://github.com/iterative/dvc/issues/2135

Reply:
also, to give a bit more context about 4:

at my previous job I saw a code like this to read the model from S3

```
contents = bucket.get_key(file_name, validate=False).set_contents_from_string(contents)
contents = zlib.decompress(contents, 15 + 32)
pickle.loads(contents)
```

or something similar.

Is it easy to rewrite it to file? Or it's easier to byte stream?

Reply:
1. Will live that for that ticket.
2. We decided from the start to do non-streaming version first, when add streaming if people will need that.
3. Don't get how copy or any semantics apply here. We pull data from remote to local cache and open from there.
4. We can sure return non-seekable file object, should not be an issue.

Reply:
People can use it like:
```python
with dvc.api.open('path/artefact.ext', repo='https://github.com/user/repo') as f:
    contents = f.read()
# ... same code here
```

Reply:
2. yes, it's not about streaming/non-streaming - that question was about a DVC-file that has multiple outputs and we need only one of them. Will it download all the outputs just to read one of them?

Reply:
actually, I checked it - it pulls a single output and reads it from cache.

Reply:
3. same - I was confused because `dvc pull` is instantiating data in the workspace. cloud.pull is just downloading it. Disregard this.